nexus/src/authz/context.rs

```diff
@@ -0,0 +1,228 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Guts of the authorization subsystem
+
+use super::actor::AnyActor;
+use super::roles::RoleSet;
+use crate::authn;
+use crate::authz::oso_generic;
+use crate::authz::Action;
+use crate::authz::AuthzResource;
+use crate::context::OpContext;
+use crate::db::DataStore;
+use omicron_common::api::external::Error;
+use oso::Oso;
+use std::sync::Arc;
+
+/// Server-wide authorization context
+pub struct Authz {
+    oso: Oso,
+}
+
+impl Authz {
+    /// Construct an authorization context
+    ///
+    /// # Panics
+    ///
+    /// This function panics if we could not load the compiled-in Polar
+    /// configuration.  That should be impossible outside of development.
+    pub fn new() -> Authz {
+        let oso = oso_generic::make_omicron_oso().expect("initializing Oso");
+        Authz { oso }
+    }
+}
+
+/// Operation-specific authorization context
+///
+/// This is the primary external interface for the authorization subsystem,
+/// through which Nexus at-large makes authorization checks.  This is almost
+/// always done through [`OpContext::authorize()`].
+pub struct Context {
+    authn: Arc<authn::Context>,
+    authz: Arc<Authz>,
+    datastore: Arc<DataStore>,
+}
+
+impl Context {
+    pub fn new(
+        authn: Arc<authn::Context>,
+        authz: Arc<Authz>,
+        datastore: Arc<DataStore>,
+    ) -> Context {
+        Context { authn, authz, datastore }
+    }
+
+    /// Check whether the actor performing this request is authorized for
+    /// `action` on `resource`.
+    pub async fn authorize<Resource>(
+        &self,
+        opctx: &OpContext,
+        action: Action,
+        resource: Resource,
+    ) -> Result<(), Error>
+    where
+        Resource: oso::ToPolar + AuthzResource,
+    {
+        // TODO-security For Action::Read (and any other "read" action),
+        // this should return NotFound rather than Forbidden.  But we cannot
+        // construct the appropriate NotFound here without more information:
+        // the resource type and how it was looked up.  In practice, it's
+        // quite possible that all such cases should really use query
+        // filtering instead of an explicit is_allowed() check, in which
+        // case we could safely assume Forbidden here.
+        //
+        // Alternatively, we could let the caller produce the appropriate
+        // "NotFound", but it would add a lot of boilerplate to a lot of
+        // callers if we didn't return api::external::Error here.
+        let mut roles = RoleSet::new();
+        resource
+            .fetch_all_related_roles_for_user(
+                opctx,
+                &self.datastore,
+                &self.authn,
+                &mut roles,
+            )
+            .await?;
```

Comment:
This is probably a basic thing about `Box`, but why do you have to use a mutable reference instead of returning `RoleSet` from here? Would a returned `RoleSet` have too short a lifetime?

Reply:
Good question.  The reason I had this accept a `&mut RoleSet` rather than returning a `RoleSet` is that we're going to call this function a bunch of times recursively.  These recursive calls can all store their results into the same data structure.  The alternative would involve allocating a new `RoleSet` in each call and then merging it after each child recursive call.  That'd be fine, but it's more code and more work at runtime.